firefox-ios/Client/Frontend/Browser/WebView/PullRefreshView.swift

```diff
@@ -0,0 +1,323 @@
+        }
+    }
+
+    private func setupSubviews() {
+        addSubviews(progressContainerView, progressView)
+        let shrinkFactor = computeShrinkingFactor()
+        let progressContainerViewPadding = UX.progressViewPadding * shrinkFactor
+        let progressContainerViewSize = UX.progressViewAnimatedBackgroundSize * shrinkFactor
+        NSLayoutConstraint.activate([
+            progressContainerView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -progressContainerViewPadding),
+            progressContainerView.centerXAnchor.constraint(equalTo: centerXAnchor),
+            progressContainerView.heightAnchor.constraint(equalToConstant: progressContainerViewSize),
+            progressContainerView.widthAnchor.constraint(equalToConstant: progressContainerViewSize),
+
+            progressView.centerYAnchor.constraint(equalTo: progressContainerView.centerYAnchor),
+            progressView.centerXAnchor.constraint(equalTo: centerXAnchor),
+            progressView.heightAnchor.constraint(equalToConstant: UX.progressViewSize * shrinkFactor),
+            progressView.widthAnchor.constraint(equalToConstant: UX.progressViewSize * shrinkFactor)
+        ])
+    }
+
+    /// Starts observing the `UIScrollView` content offset so the refresh can react to content change and show
+    /// the correct state of the pull to refresh view.
+    func startObservingContentScroll() {
+        obeserveTicket = scrollView?.observe(\.contentOffset) { [weak self] _, _ in
+            guard let scrollView = self?.scrollView, scrollView.isDragging
+            else {
+                guard let refreshHasFocus = self?.refreshIconHasFocus, refreshHasFocus else { return }
+                self?.refreshIconHasFocus = false
+                self?.easterEggGif?.removeFromSuperview()
+                self?.easterEggGif = nil
+                self?.obeserveTicket?.invalidate()
+                self?.triggerReloadAnimation()
+                return
+            }
+            let threshold = (self?.computeShrinkingFactor() ?? 1.0) * UX.blinkProgressViewStandardThreshold
+            if scrollView.contentOffset.y < -threshold {
+                self?.blinkBackgroundProgressViewIfNeeded()
+                DispatchQueue.main.asyncAfter(deadline: .now() + UX.easterEggDelayInSeconds) {
+                    self?.showEasterEgg()
+                }
+            } else if scrollView.contentOffset.y != 0.0 {
+                // This check prevents progressView re blink when scrolling the pull refresh before the web view is loaded
+                self?.restoreBackgroundProgressViewIfNeeded()
+                let rotationAngle = -(scrollView.contentOffset.y / (self?.frame.height ?? 0.0)) * .pi * 2
+                UIView.animate(withDuration: 0.1) {
+                    self?.progressView.transform = CGAffineTransform(rotationAngle: rotationAngle * 2.0)
+                }
+            }
+        }
```

Comment:
nitpick: Maybe we can add a few spaces between code blocks for readability? 

```suggestion
                guard let refreshHasFocus = self?.refreshIconHasFocus, refreshHasFocus else { return }
                self?.refreshIconHasFocus = false
                self?.easterEggGif?.removeFromSuperview()
                self?.easterEggGif = nil
                self?.obeserveTicket?.invalidate()
                self?.triggerReloadAnimation()
                return
            }
            
            let threshold = (self?.computeShrinkingFactor() ?? 1.0) * UX.blinkProgressViewStandardThreshold
            
            if scrollView.contentOffset.y < -threshold {
                self?.blinkBackgroundProgressViewIfNeeded()
                
                DispatchQueue.main.asyncAfter(deadline: .now() + UX.easterEggDelayInSeconds) {
                    self?.showEasterEgg()
                }
            } else if scrollView.contentOffset.y != 0.0 {
                // This check prevents progressView re blink when scrolling the pull refresh before the web view is loaded
                self?.restoreBackgroundProgressViewIfNeeded()
                let rotationAngle = -(scrollView.contentOffset.y / (self?.frame.height ?? 0.0)) * .pi * 2
                
                UIView.animate(withDuration: 0.1) {
                    self?.progressView.transform = CGAffineTransform(rotationAngle: rotationAngle * 2.0)
                }
            }
        }
```

Reply:
Ok thanks so much @ih-codes i'll add those